course/viewbase.py

```diff
@@ -57,31 +57,43 @@ def get_resource_objects(self):
             self.instance = instance
             self.course = self.instance.course
             self.content = CachedContent(self.instance)
+            self.user_in_course = None
+            if user.is_authenticated:
+                self.user_in_course = self.instance.get_enrollment_for(user)
             self.is_student = self.instance.is_student(user)
             self.is_assistant = self.instance.is_assistant(user)
             self.is_teacher = self.course.is_teacher(user)
             self.is_course_staff = self.is_teacher or self.is_assistant
             self.get_taggings = lambda: CachedStudent(instance, user.id).data['tag_slugs']
 
             self.note(
-                "course", "instance", "content", "is_student", "is_assistant",
+                "course", "instance", "content", "user_in_course", "is_student", "is_assistant",
                 "is_teacher", "is_course_staff", "get_taggings",
             )
 
-            # Apply course instance language.
+            # Try to find a language that is defined for this course instance
+            # and apply it
             if self.instance.language:
-                lang = self.instance.language
-                if lang.startswith("|"):
-                    active = get_language()
-                    if "-" in active:
-                        active = active.split("-")[0]
-                    if "|" + active + "|" in lang:
-                        translation.activate(active)
-                    else:
-                        fallback = lang[1:lang.find("|", 1)]
-                        translation.activate(fallback)
+                instance_languages = self.instance.language
+
+                languages = []
+                if self.user_in_course and self.user_in_course.language:
+                    languages.append(self.user_in_course.language)
+                if user.is_authenticated and user.userprofile.language:
+                    languages.append(user.userprofile.language)
+                languages.append(get_language())
+
+                for lang in languages:
+                    lang = lang.split("-")[0]
+                    if lang in instance_languages:
+                        language = lang
+                        break
                 else:
-                    translation.activate(lang)
+                    if instance_languages.startswith("|"):
+                        language = instance_languages[1:instance_languages.find("|", 1)]
```

Comment:
This is good. Alternative would be using `split`, but both are relatively readable, so this is fine. But just for your info:

```
>>> timeit("lg = lgs[1:].partition('|')[0]", "lgs = '|en|fi|sv|ru|'")
0.15018817299278453
>>> timeit("lg = lgs[1:lgs.find('|', 1)]", "lgs = '|en|fi|sv|ru|'")
0.14492674503708258
>>> timeit("lg = lgs.split('|', 2)[1]", "lgs = '|en|fi|sv|ru|'")
0.11502113298047334
```

Reply:
Btw. #620 will introduce `instance.default_language`, which could be used here, once that is merged.

Reply:
Thanks!
> #620 will introduce instance.default_language, which could be used here, once that is merged.

Nice, I'll use that when it's merged.

Reply:
#620 is now merged, but it is not yet in the master branch since it is part of the hotfixes that we want to deploy to production in August.

Reply:
Apparently, we didn't need that here after all.